File: e2e/java/src/test/java/dev/kreuzberg/e2e/MetadataTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/** E2e tests for category: metadata. */
class MetadataTest {
    @Test
    void testMetadataArticleTimes() throws Exception {
        // Extracts article:published_time, modified_time, author, section, and tags
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'article.published_time' not available on result type
        // skipped: field 'article.modified_time' not available on result type
        // skipped: field 'article.author' not available on result type
        // skipped: field 'article.section' not available on result type
        // skipped: field 'article.tags.length' not available on result type
    }

    @Test
    void testMetadataFavicons() throws Exception {
        // Extracts favicon link tags including apple-touch-icon
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'favicons.length' not available on result type
        // skipped: field 'favicons[].apple_touch' not available on result type
    }

    @Test
    void testMetadataHeadings() throws Exception {
        // Extracts heading hierarchy (h1-h6) from HTML page
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'headings.h1.length' not available on result type
        // skipped: field 'headings.h1[0].text' not available on result type
        // skipped: field 'headings.length' not available on result type
    }

    @Test
    void testMetadataHreflang() throws Exception {
        // Extracts hreflang alternate link tags
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'hreflang.length' not available on result type
        // skipped: field 'hreflang[].lang' not available on result type
    }

    @Test
    void testMetadataKeywordsAuthor() throws Exception {
        // Extracts keywords, author, viewport, generator, theme-color, robots, lang, dir metadata
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertEquals("Comprehensive Metadata Test Page", result.metadata().orElseThrow().title().orElse(""));
        assertFalse(result.metadata().orElseThrow().canonicalUrl().orElse("").isEmpty(), "expected non-empty value");
        assertFalse(result.metadata().orElseThrow().keywords().isEmpty(), "expected non-empty value");
        assertTrue(result.metadata().orElseThrow().keywords().contains("rust"), "expected to contain: " + "rust");
        assertEquals("Jane Developer", result.metadata().orElseThrow().author());
        assertFalse(result.metadata().orElseThrow().viewport().isEmpty(), "expected non-empty value");
        assertEquals("kreuzcrawl/1.0", result.metadata().orElseThrow().generator());
        assertEquals("#ff6600", result.metadata().orElseThrow().themeColor());
        assertEquals("index, follow", result.metadata().orElseThrow().robots());
        assertEquals("en", result.metadata().orElseThrow().lang());
        assertEquals("ltr", result.metadata().orElseThrow().dir());
    }

    @Test
    void testMetadataOgVideoAudio() throws Exception {
        // Extracts og:video, og:audio, and og:locale:alternate metadata
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'og.video' not available on result type
        // skipped: field 'og.audio' not available on result type
        // skipped: field 'og.locale_alternate.length' not available on result type
    }

    @Test
    void testMetadataResponseHeaders() throws Exception {
        // Extracts response metadata from HTTP headers (etag, server, content-language)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'response_headers.etag' not available on result type
        // skipped: field 'response_headers.last_modified' not available on result type
        // skipped: field 'response_headers.server' not available on result type
        // skipped: field 'response_headers.content_language' not available on result type
    }

    @Test
    void testMetadataWordCount() throws Exception {
        // Computes word count from visible page text
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        // skipped: field 'computed.word_count' not available on result type
        // skipped: field 'computed.word_count' not available on result type
    }

}

File: e2e/elixir/test/metadata_test.exs
# E2e tests for category: metadata
defmodule E2e.MetadataTest do
  use ExUnit.Case, async: true

  describe "metadata_article_times" do
    test "Extracts article:published_time, modified_time, author, section, and tags" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(result.article.published_time) == "2024-01-15T10:00:00Z"
      assert String.trim(result.article.modified_time) == "2024-06-20T14:30:00Z"
      assert String.trim(result.article.author) == "Jane Developer"
      assert String.trim(result.article.section) == "Technology"
      assert String.trim(length(result.article.tags)) == 3
    end
  end

  describe "metadata_favicons" do
    test "Extracts favicon link tags including apple-touch-icon" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.favicons)) == 5
      assert result.favicons.get("").apple_touch != ""
    end
  end

  describe "metadata_headings" do
    test "Extracts heading hierarchy (h1-h6) from HTML page" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.headings.h1)) == 1
      assert String.trim(result.headings.h1.get("0").text) == "Primary Heading"
      assert String.trim(length(result.headings)) == 8
    end
  end

  describe "metadata_hreflang" do
    test "Extracts hreflang alternate link tags" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.hreflang)) == 4
      assert String.contains?(result.hreflang.get("").lang, "en")
    end
  end

  describe "metadata_keywords_author" do
    test "Extracts keywords, author, viewport, generator, theme-color, robots, lang, dir metadata" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(result.metadata.title) == "Comprehensive Metadata Test Page"
      assert result.metadata.canonical_url != ""
      assert result.metadata.keywords != ""
      assert String.contains?(result.metadata.keywords, "rust")
      assert String.trim(result.metadata.author) == "Jane Developer"
      assert result.metadata.viewport != ""
      assert String.trim(result.metadata.generator) == "kreuzcrawl/1.0"
      assert String.trim(result.metadata.theme_color) == "\#ff6600"
      assert String.trim(result.metadata.robots) == "index, follow"
      assert String.trim(result.metadata.lang) == "en"
      assert String.trim(result.metadata.dir) == "ltr"
    end
  end

  describe "metadata_og_video_audio" do
    test "Extracts og:video, og:audio, and og:locale:alternate metadata" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert String.trim(result.og.video) == "https://example.com/video.mp4"
      assert String.trim(result.og.audio) == "https://example.com/audio.mp3"
      assert String.trim(length(result.og.locale_alternate)) == 2
    end
  end

  describe "metadata_response_headers" do
    test "Extracts response metadata from HTTP headers (etag, server, content-language)" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert result.response_headers.etag != ""
      assert result.response_headers.last_modified != ""
      assert String.contains?(result.response_headers.server, "nginx")
      assert String.trim(result.response_headers.content_language) == "en-US"
    end
  end

  describe "metadata_word_count" do
    test "Computes word count from visible page text" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
      assert result.computed.word_count > 99
      assert result.computed.word_count < 301
    end
  end
end
